### PR TITLE
feat(dataset): add subtask support

### PR DESCRIPTION
**Pull request by jadechoghari**
## Title

feat(dataset): add subtask support

### CODE REVIEW

File: src/lerobot/utils/constants.py [hunk 1]
```diff
@@ -26,6 +26,9 @@
 OBS_LANGUAGE = OBS_STR + ".language"
 OBS_LANGUAGE_TOKENS = OBS_LANGUAGE + ".tokens"
 OBS_LANGUAGE_ATTENTION_MASK = OBS_LANGUAGE + ".attention_mask"
+OBS_LANGUAGE_SUBTASK = OBS_STR + ".subtask"
+OBS_LANGUAGE_SUBTASK_TOKENS = OBS_LANGUAGE_SUBTASK + ".tokens"
+OBS_LANGUAGE_SUBTASK_ATTENTION_MASK = OBS_LANGUAGE_SUBTASK_TOKENS + ".attention_mask"
```
Comment (michel-aractingi) [inline src/lerobot/utils/constants.py]: Is the attn mask supposed to be taken from `OBS_LANGUAGE_SUBTASK` or `OBS_LANGUAGE_SUBTASK_TOKENS` .
I think  its supposed to return `observation.subtask.attention_mask` not `observation.subtask.tokens.attention_mask` 
```suggestion
OBS_LANGUAGE_SUBTASK_ATTENTION_MASK = OBS_LANGUAGE_SUBTASK + ".attention_mask"
```

File: src/lerobot/policies/sarm/processor_sarm.py [hunk 2]
```diff
@@ -382,8 +479,144 @@ def _compute_batch_targets(
...
+        For the new per-frame format, we need to estimate how far along we are
+        in the current subtask by finding where it starts and ends in this episode.
+        """
+        # Find subtask boundaries by scanning episode
+        # For efficiency, we use a simple heuristic: estimate based on position
+        # within a local window of same-stage frames
+
+        # First, find the start of current subtask (scan backwards)
+        subtask_start = abs_idx
+        for idx in range(abs_idx - 1, ep_start - 1, -1):
```
Comment (michel-aractingi) [inline src/lerobot/policies/sarm/processor_sarm.py]: would this function be too slow to scan over episodes twice? can we do caching here?

File: src/lerobot/data_processing/sarm_annotations/subtask_annotation.py [hunk 3]
```diff
@@ -742,10 +742,120 @@ def visualize_annotations(
...
+
+    dataset_path = dataset.root
+
+    # 1. Create meta/subtasks.parquet in original dataset
+    subtasks_df = pd.DataFrame({"subtask_index": range(len(subtask_list))}, index=subtask_list)
+    subtasks_df.index.name = None  # Remove index name
+    subtasks_path = dataset_path / DEFAULT_SUBTASKS_PATH
+    subtasks_path.parent.mkdir(parents=True, exist_ok=True)
+    subtasks_df.to_parquet(subtasks_path, engine="pyarrow", compression="snappy")
+    print(f"Saved subtasks mapping to {subtasks_path}")
```
Comment (michel-aractingi) [inline src/lerobot/data_processing/sarm_annotations/subtask_annotation.py]: can you use logging instead of prints in this entire script

File: src/lerobot/policies/sarm/processor_sarm.py (hunk 2 above)
Comment (jadechoghari) [reply src/lerobot/policies/sarm/processor_sarm.py]: SARM subtask/high level task will be handled ina. future PR.

File: src/lerobot/data_processing/sarm_annotations/subtask_annotation.py (hunk 3 above)
Comment (jadechoghari) [reply src/lerobot/data_processing/sarm_annotations/subtask_annotation.py]: SARM subtask/high level task will be handled in a future PR.

File: src/lerobot/datasets/utils.py [hunk 4]
```diff
@@ -353,6 +354,14 @@ def load_tasks(local_dir: Path) -> pandas.DataFrame:
     return tasks
 
 
+def load_subtasks(local_dir: Path) -> pandas.DataFrame | None:
+    """Load subtasks from subtasks.parquet if it exists."""
+    subtasks_path = local_dir / DEFAULT_SUBTASKS_PATH
+    if subtasks_path.exists():
+        return pd.read_parquet(subtasks_path)
+    return None
```
[the remark this answers is not in the document]
Comment (jadechoghari) [reply src/lerobot/datasets/utils.py:362]: we never used write_subtasks so far, we'll add it, once we plan to use such function

File: src/lerobot/processor/tokenizer_processor.py [hunk 5]
```diff
@@ -176,6 +204,24 @@ def observation(self, observation: RobotObservation) -> RobotObservation:
...
+                tokenized_subtask = {
+                    k: v.to(target_device) if isinstance(v, torch.Tensor) else v
+                    for k, v in tokenized_subtask.items()
+                }
+
+            # Add tokenized subtask to the observation
+            new_observation[OBS_LANGUAGE_SUBTASK_TOKENS] = tokenized_subtask["input_ids"]
+            new_observation[OBS_LANGUAGE_SUBTASK_ATTENTION_MASK] = tokenized_subtask["attention_mask"].to(
+                dtype=torch.bool
+            )
```
[the remark this answers is not in the document]
Comment (jadechoghari) [reply src/lerobot/processor/tokenizer_processor.py:223]: tests are added

File: src/lerobot/datasets/lerobot_dataset.py [hunk 6]
```diff
@@ -1075,6 +1078,12 @@ def __getitem__(self, idx) -> dict:
         # Add task as a string
         task_idx = item["task_index"].item()
         item["task"] = self.meta.tasks.iloc[task_idx].name
+
+        # add subtask information if available
+        if "subtask_index" in self.features and self.meta.subtasks is not None:
+            subtask_idx = item["subtask_index"].item()
+            item["subtask"] = self.meta.subtasks.iloc[subtask_idx].name
+
```
[the remark this answers is not in the document]
Comment (jadechoghari) [reply src/lerobot/datasets/lerobot_dataset.py:1086]: tests are added

File: docs/source/dataset_subtask.mdx [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+When subtasks are available, each sample includes:
+
+| Key             | Type           | Description                          |
+| --------------- | -------------- | ------------------------------------ |
+| `subtask_index` | `torch.Tensor` | Integer index of the current subtask |
+| `subtask`       | `str`          | Natural language subtask description |
+
+## Related Resources
+
+- [SARM Paper](https://arxiv.org/pdf/2509.25358) - Stage-Aware Reward Modeling for Long Horizon Robot Manipulation
```
Comment (pkooij) [inline docs/source/dataset_subtask.mdx:276]: Maybe add hi-robot paper as reference as well

Comment (jadechoghari) [reply docs/source/dataset_subtask.mdx:276]: yes will do in a follow up PR
